src/cockpit/samples.py

```diff
@@ -95,47 +97,67 @@ def sample(self, samples: Samples) -> None:
-        for index in range(1, 2 ** 32):
-            sensor_path = f'/sys/class/hwmon/hwmon{hwmonid}/temp{index}_input'
-            if not os.path.exists(sensor_path):
-                break
-
-            with open(f'/sys/class/hwmon/hwmon{hwmonid}/temp{index}_label') as f:
-                label = f.read().strip()
-            if label:
-                # only sample CPU Temperature in atk0110
-                if label != 'CPU Temperature' and name == 'atk0110':
-                    continue
-                # ignore Tctl on AMD devices
-                if label == 'Tctl':
+    @staticmethod
+    def detect_cpu_sensors(dir_fd: int) -> Iterable[str]:
+        # Read the name file to decide what to do with this directory
+        try:
+            with Handle.open('name', os.O_RDONLY, dir_fd=dir_fd) as fd:
+                name = os.read(fd, 1024).decode().strip()
+        except FileNotFoundError:
+            return
+
+        if name == 'atk0110':
+            # only sample 'CPU Temperature' in atk0110
+            predicate = (lambda label: label == 'CPU Temperature')
+        elif name == 'cpu_thermal':
+            # labels are not used on ARM
+            predicate = None
+        elif name == 'coretemp':
+            # accept all labels on Intel
+            predicate = None
+        elif name in ['k8temp', 'k10temp']:
+            # ignore Tctl on AMD devices
+            predicate = (lambda label: label != 'Tctl')
+        else:
+            # Not a CPU sensor
+            return
+
+        # Now scan the directory for inputs
+        for input_filename in os.listdir(dir_fd):
+            if not input_filename.endswith('_input'):
+                continue
+
+            if predicate:
+                # We need to check the label
+                try:
+                    label_filename = input_filename.replace('_input', '_label')
+                    with Handle.open(label_filename, os.O_RDONLY, dir_fd=dir_fd) as fd:
+                        label = os.read(fd, 1024).decode().strip()
+                except FileNotFoundError:
                     continue
-            else:
-                # labels are not used on ARM
-                if name != 'cpu_thermal':
+
+                if not predicate(label):
                     continue
 
-            self.sensors.append(sensor_path)
+            yield input_filename
 
-    def sample(self, samples: Samples) -> None:
-        cpu_names = ['coretemp', 'cpu_thermal', 'k8temp', 'k10temp', 'atk0110']
+    @staticmethod
+    def scan_sensors() -> Iterable[str]:
+        with Handle.open(HWMON_PATH, os.O_RDONLY | os.O_DIRECTORY) as top_fd:
+            for hwmon_name in os.listdir(top_fd):
+                with Handle.open(hwmon_name, os.O_RDONLY | os.O_DIRECTORY, dir_fd=top_fd) as subdir_fd:
+                    for sensor in CPUTemperatureSampler.detect_cpu_sensors(subdir_fd):
+                        yield f'{HWMON_PATH}/{hwmon_name}/{sensor}'
```

Comment:
This path mangling! :P

Reply:
ya, what can you do? :)  We could store a list of fds to read from....